kernel/src/dynamic_process_loading.rs

```diff
@@ -0,0 +1,1242 @@
+                        }
+                        // Otherwise we let the client know we are done with the setup, and we are ready to write the app to flash
+                        None | Some(PaddingRequirement::PrePad) => {
+                            self.state.set(State::FirstWrite);
+                            self.client.map(|client| {
+                                client.setup_done();
+                            });
+                        }
+                    };
+                    Ok(app_length)
+                }
+                Err(_err) => {
+                    self.reset_process_loading_metadata(); // reset the state to None because we did not find any available address for this app
+                    Err(ErrorCode::FAIL)
+                }
+            }
+        } else {
+            Err(ErrorCode::BUSY) // this means the kernel is busy doing some other operation already.
+        }
+    }
+
+    fn write_app_data(
+        &self,
+        buffer: SubSliceMut<'static, u8>,
+        offset: usize,
+        _processid: ProcessId,
+    ) -> Result<(), ErrorCode> {
+        match self.state.get() {
+            Some(State::FirstWrite) => {
+                if offset != self.next_offset.get() {
+                    // self.next_offset should be 0 at this stage.
+                    self.reset_process_loading_metadata(); // clear out new app size, start address, current offset and state
+                    return Err(ErrorCode::INVAL); // this means the app is trying to write to an address that is not the
+                                                  // start address assigned to it after setup. Return an Invalid error.
+                }
+                let res = self.write(buffer, offset);
+                match res {
+                    Ok(()) => Ok(()),
+                    Err(e) => {
+                        // if we fail here, let us erase the app we just wrote
+                        self.state.set(State::PaddingWrite);
+                        let _ = self.write_padding_app(
+                            self.new_app_length.get(),
+                            self.new_app_start_addr.get(),
+                        );
+                        self.reset_process_loading_metadata(); // clear out new app size, start address, current offset and state
+                        Err(e)
+                    }
+                }
+            }
+            Some(State::AppWrite) => {
+                // the app is trying to write to an address that is not the immediate continuation of where
+                // the previous write left off. Return an Invalid error.
+                if offset != self.next_offset.get() {
+                    // check if the app is trying to write to the correct offset
+                    self.reset_process_loading_metadata(); // clear out new app size, start address, current offset and state
+                    return Err(ErrorCode::INVAL);
+                }
```

Comment:
Oh, I see. But what if the loader is doing compression where it doesn't need to write every byte?

Reply:
I did not think of that